src/mongoc/mongoc-topology-description.c

```diff
@@ -0,0 +1,996 @@
+      return 0;
+   }
+
+   id = ++topology->max_server_id;
+
+   description = bson_malloc0(sizeof *description);
+   mongoc_server_description_init(description, server, id);
+
+   mongoc_set_add(topology->servers, id, description);
+
+   if (topology->cb.add) {
+      topology->cb.add(description);
+   }
+
+   return id;
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _mongoc_update_rs_from_primary --
+ *
+ *       First, determine that this is really the primary:
+ *          -If this node isn't in the cluster, do nothing.
+ *          -If the cluster's set name is null, set it to node's set name.
+ *           Otherwise if the cluster's set name is different from node's,
+ *           we found a rogue primary, so remove it from the cluster and
+ *           check the cluster for a primary, then return.
+ *          -If any of the members of cluster reports an address different
+ *           from node's, node cannot be the primary.
+ *       Now that we know this is the primary:
+ *          -If any hosts, passives, or arbiters in node's description aren't
+ *           in the cluster, add them as UNKNOWN servers and begin monitoring.
+ *          -If the cluster has any servers that aren't in node's description
+ *           remove them and stop monitoring.
+ *       Finally, check the cluster for the new primary.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       Changes to the cluster, possible removal of cluster nodes.
+ *
+ *--------------------------------------------------------------------------
+ */
+static void
+_mongoc_topology_description_update_rs_from_primary (mongoc_topology_description_t *topology,
+                                                     mongoc_server_description_t   *server)
+{
+   mongoc_server_description_t *server_iter;
+   mongoc_set_t *set = topology->servers;
+   bson_iter_t member_iter;
+   const bson_t *rs_members[3];
+   const char *member_name;
+   int i;
+
+   if (!_mongoc_topology_description_has_server(topology, server->connection_address)) return;
+
+   /*
+    * 'Server' can only be the primary if it has the right rs name.
+    */
+
+   if (!topology->set_name) {
+      int len = strlen(server->set_name) + 1;
+      topology->set_name = bson_malloc (len);
+      memcpy (topology->set_name, server->set_name, len);
+   }
+   else if (topology->set_name != server->set_name) {
+      _mongoc_topology_description_remove_server(topology, server);
+      _mongoc_topology_description_check_if_has_primary(topology, server);
+      return;
+   }
+
+   /*
+    *'Server' is the primary! Invalidate other primaries if found.
+    */
+
+   for (i = 0; i < set->items_len; i++) {
+      server_iter = set->items[i].item;
+      if (server_iter->id != server->id &&
+          server_iter->type == MONGOC_SERVER_RS_PRIMARY) {
+         mongoc_server_description_set_state(server, MONGOC_SERVER_UNKNOWN);
+         // TODO SDAM set other states to defaults?
```

Comment:
No, the primary changed but the other servers' states are probably unchanged, just wait for the next scan to re-check them.